### ISSUE TITLE
variable type deduction does not propagate into inner functions 

### ISSUE THREAD

**Issue by erichang-bcad** [python/mypy#9988]
**Bug Report**

when mypy deduces variable types (from if statement checks, for example), the result of the deduction doesn't apply to inner functions

**To Reproduce**

Create a simple python file:
```python
from __future__ import annotations

def test(foo: int | None) -> None:
    if foo is not None:
        def bar() -> int:
            return -foo + 1
        print(f"bar returns: {bar()}")
    else:
        print("foo is None")


if __name__ == '__main__':
    test(3)
    test(None)

```

then run mypy on this file

**Expected Behavior**

mypy should show no errors, because the inner function ``bar`` should know that variable ``foo`` is an ``int``.

**Actual Behavior**

mypy reports the error:
```
test_mypy.py:7: error: Unsupported operand type for unary - ("Optional[int]")
Found 1 error in 1 file (checked 1 source file)
```

**Your Environment**

- Mypy version used: 0.800
- Mypy command-line flags: None 
- Mypy configuration options from `mypy.ini` (and other config files): None
- Python version used: 3.9
- Operating system and version: Arch Linux

**Comment by hauntsaninja** [comment]
https://github.com/python/mypy/issues/2608#issuecomment-727042292
